modules/tests/shared/src/test/scala/io/circe/DecoderSuite.scala

```diff
@@ -779,4 +780,55 @@ class DecoderSuite extends CirceMunitSuite with LargeNumberDecoderTestsMunit {
     assert(result.isInvalid)
     assertEquals(result.swap.toOption.map(_.size), Some(2))
   }
+
+  checkAll("Defer[Decoder]", DeferTests[Decoder].defer[MiniInt])
+
+  test("Decoder.recursive should prevent undesirable grown in the number of instances created") {
+    var counter = 0
+    implicit def uglyListDecoder[A: Decoder]: Decoder[List[A]] = {
+      counter += 1
+      Decoder.recursive[List[A]] { implicit recurse =>
+        Decoder.instance { c =>
+          (c.downField("car").as[A], c.downField("cdr").as[Option[List[A]]]).mapN(_ :: _.getOrElse(Nil))
+        }
+      }
+    }
+    assertEquals(
+      Json
+        .obj(
+          "car" := 0,
+          "cdr" := Json.obj(
+            "car" := 1,
+            "cdr" := Json.obj(
+              "car" := 2,
+              "cdr" := Json.obj(
+                "car" := 3
+              )
+            )
+          )
+        )
+        .as[List[Int]],
+      (0 :: 1 :: 2 :: 3 :: Nil).asRight
+    )
+    // Without `Decoder.recursive`, this should create 5 instances of a `Decoder[List[Int]]`
+    // (1 for each instance, and 1 that gets created but not called because because the last "cdr" field is missing)
+    assertEquals(counter, 1)
+  }
+
+  test("Decoder.recursive should be stack safe") {
+    // Pretty much the same as the one in the other test, we just don't need to count for this one
+    implicit def uglyListDecoder[A: Decoder]: Decoder[List[A]] =
+      Decoder.recursive[List[A]] { implicit recurse =>
+        Decoder.instance { c =>
+          (c.downField("car").as[A], c.downField("cdr").as[Option[List[A]]]).mapN(_ :: _.getOrElse(Nil))
```

Comment:
I don't think this is stack safe, and it looks like the tests fail.

So, recurse is doing to be used at `c.downField("cdr").as[Option[List[A]]]` but that's nested inside of a `mapN`. While nesteding `Defer` is stack safe in the construction you have here, composing with map2, which is what is happening here, isn't. Something like: `lazy val d = map2(a, defer(d))(_ :: _)`, `map2` would need to somehow unroll that loop. It's not easy to see how that could happen.

So, I think the stack depth will be the same depth as the object graph.

cats-parse has the same issue on parsing (which is the same basic typeclass as Decoder).

Reply:
Something like this?

> Note: while a `Decoder` written using `Decoder.recursive` can prevent unneeded creation of `Decoder` instances when recursing, the resulting `Decoder` cannot be guaranteed to be stack-safe.

Reply:
yeah, something like that seems right.